cpp/TUTORIAL.md

```diff
@@ -0,0 +1,337 @@
+To minimize the state which is transferred between contracts and stored in the blockchain, TVM does not have random access memory. The memory might be emulated by dictionaries, however, it’s quite expensive and often not practical taking into account limited gas supply. Thus a typical smart contract in C++ avoid use memory. Moreover, currently all the functions have to be inlined. That’s why `-O3` optimization and LTO are essential and it’s the pipeline `tvm-build++` implements.
+
+## Environment setup
+Prior to start developing a contract, we configure `PATH` to add all the tools we need:
+```
+export PATH=TON-Compiler-build/bin:TVM-linker-source/tvm_linker/target/<debug or release>:TVM-linker-source/tonlabs-cli/target/<debug or release>:$PATH
+```
+`tvm-build++` needs additional theatment:
+```
+export TVM_LINKER=TVM-linker-source/tvm_linker/target/<debug or release>/tvm_linker #path to tvm_linker tool.
+export TVM_INCLUDE_PATH=TON-Compiler-source/stdlib #path to the folder containing cpp-sdk directory.
+export TVM_LIBRARY_PATH=TON-Compiler-source/stdlib #path to stdlib_cpp.tvm
+export LD_LIBRARY_PATH=TVM-linker-source/tonlabs-cli/target/<debug or release>/ ton-client-rs-<hash>/out/libton_client.so:$LD_LIBRARY_PATH #path to the tonlabs-cli library
+```
+
+## Hello, world!
+A typical smart contract consists of two files:
+*	contract interface description in a header file
+*	contract implementation in a cpp-file
+A contract might consist of more files if it's needed, but since the contract tend to be small, two files are often enough.
+Let's start developing our first, `Hello, world!` contract by describing its interface.
+
+### Describing a contract interface
+A contract interface consists of three parts:
+*	Public methods declarations.
+*	Persistent data.
+*	Events.
+Public methods are functions that receive messages in the blockchain. There is a dedicated method called `constructor` which is called upon the contract deploy. The constructor is a must otherwise the contract can not be deployed. The constructor, as well as other public methods, must only take arguments of types that are listed below:
+*	`int_t<N>` - N-bits wide signed integer, 0 < N < 256.
+*	`uint_t<N>` - N-bits wide unsigned integer, 0 < N < 257.
+*	`MsgAddress` – message address which work for both internal or external messages.
+*	`MsgAddressInt` – an internal message address.
+*	`MsgAddressExt` – an external message address.
+*	`dict_array<T, 32>` - a map representing an "array" with up to 32 elements of type `T` (indexed from 0 to 31). `T` must also belong to this list.
+*	`dict_map<KeyT, ValueT>` - a map. Both `KeyT` and `ValueT` must also belong to the list.
```

Comment:
Actually, KeyT must be `uint_t<N>` strictly (in dict_map).

Reply:
Fixed